### PR TITLE
feat: Added Functions (UDF) Resource & Datasource

### LINKED ISSUE #532: Stored procedures and functions
**Is your feature request related to a problem? Please describe.**

Is there a plan to support stored procedures and functions as resources? I see the grant for function is already here, but I cannot find function as a resource.

**Describe the solution you'd like**

Two new resources: stored procedure and user defined function.

**Describe alternatives you've considered**

.

**Additional context**

.

### PR DESCRIPTION
**Pull request by ChrisIsidora**
Implemented Resources & Datasource for User Defined Functions and made arguments in functions grant optional

## Test Plan
* [x] acceptance tests
* [x] unit tests

## References
* Closes #532 
* https://docs.snowflake.com/en/sql-reference/sql/create-function.html

### CODE REVIEW

**Comment by ChrisIsidora** [comment]
@alldoami Is it possible to make a new version after this is merged? Thanks in advance

File: pkg/datasources/functions.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		functionMap["return_type"] = functionSignatureMap["returnType"].(string)
+
+		functions = append(functions, functionMap)
+	}
+
+	d.SetId(fmt.Sprintf(`%v|%v`, databaseName, schemaName))
+	return d.Set("functions", functions)
+}
+
+func parseArguments(arguments string) (map[string]interface{}, error) {
```
Comment (alldoami) [inline pkg/datasources/functions.go:113]: This is already defined in `procedures.go`

Comment (ChrisIsidora) [reply pkg/datasources/functions.go]: @alldoami  Thanks, Removed.

**Review by alldoami — APPROVED** [review]

**Comment by Dinesh-Rupchandani** [comment]
I understand that we can now create SQL UDFs using terraform resources.
Can we get usage examples for SQL UDF like we have for python/java languages.

Thanks

**Pull request merged**